include/oneapi/dpl/pstl/hetero/algorithm_impl_hetero.h

```diff
@@ -428,23 +428,22 @@ __pattern_min_element(_ExecutionPolicy&& __exec, _Iterator __first, _Iterator __
     using _IteratorValueType = typename ::std::iterator_traits<_Iterator>::value_type;
     using _IndexValueType = ::std::make_unsigned_t<typename ::std::iterator_traits<_Iterator>::difference_type>;
     using _ReduceValueType = tuple<_IndexValueType, _IteratorValueType>;
-#ifdef _ONEDPL_DETECT_SPIRV_COMPILATION
-    using _Commutative = ::std::false_type;
-    // This operator doesn't track the lowest found index in case of equal min. or max. values. Thus, this operator is
-    // not commutative.
+    using _Commutative = ::std::true_type;
```

Comment:
I believe this is a bug.  We are hardcoding to be commutative = true, but it is not always the case in practice as it depends on `_ONEDPL_DETECT_SPIRV_COMPILATION`.  We check this in unseq_backend_sycl.h and can end up in code which does not match our operator.

Additionally, I think it is a design problem to embed this knowledge of commutativity in the kernel name by having it in the type if it might depend upon knowledge that can only be known at JIT time.  We may want to rethink it.


With this in mind, here is my tentative proposal:  
1) Add a new struct:
`_CommutativeUnknown` which is passed in place of `std::true_type` or `std::false_type` as the `_Commutative` template parameter to make it clear that this type should not be used to make actual decisions in this pattern as we do not know at host compilation time the status of this.

2) add a struct `__is_commutative_known` which determines if we can make a decision from the _Commutative value, or if we need to have special casing.  

3) add special handling in `unseq_backend_sycl.h` to use `__is_commutative_known` and if it is unknown, use a direct check of `_ONEDPL_DETECT_SPIRV_COMPILATION` and choose the appropriate code.  I believe these are the only active places where `_Commutative` is used in practice to functionally change the code, but this should be confirmed.  If there are more, we need similar handling of a third potential value for `_Commutative`.

This probably needs more thought, but I believe something like this will be needed.

Reply:
After some offline discussion, I have some updates to this proposal.  

Add a new struct which provides a matrix of commutative true / false decisions for spriv and non-spirv targets, which the host knows and can specify at host compile time.  `template<typename _SpirvCommutative, typename _NonSpirvCommutative> struct _CommutativeMatrix`

Then add a helper struct to check if commutative which has overloads for `std::true_type` and `std::false` type for backward compatibility, but also for `_CommutativeMatrix` which  uses `_ONEDPL_DETECT_SPIRV_COMPILATION` to decide commutativity at kernel compilation time to result in the correct decision.
  
The kernel name will be consistent since all the information is in the `_Commutative` type, and the choice is at kernel compilation time.

Reply:
I have went ahead and added my implementation of these new structs to address this issue and test CI.

I am looking at the best way to consolidate `_ONEDPL_DETECT_SPIRV_COMPILATION` checks preferably in the structs to cleanup up the code and have a single field in `__commutative_matrix` here.

Reply:
I have pushed the change to maintain a single `value` field in `__is_commutative_operator`.

As discussed offline, the `_ONEDPL_DETECT_SPIRV_COMPILATION` checks remain to force the non-commutative path for SPIRV targets.

Reply:
looks good.

Reply:
I like the idea overall, but I do not like that the implementation is de-facto split across 3 different files.